### ISSUE TITLE
Output of child process launched with Open3.popen3 is limited to 65536 characters

### ISSUE THREAD

**Issue by yamam** [jruby/jruby#8235]
**Environment Information**
 OpenJDK 21 or 22 
https://download.java.net/java/GA/jdk22.0.1/c7ec1332f7bb44aeba2eb341ae18aca4/8/GPL/openjdk-22.0.1_linux-x64_bin.tar.gz
https://download.java.net/java/GA/jdk21.0.2/f2283984656d49d69e91c558476027ac/13/GPL/openjdk-21.0.2_linux-x64_bin.tar.gz
```
$ ./jdk-22.0.1/bin/java -jar jruby-complete-9.4.7.0.jar -v
jruby 9.4.7.0 (3.1.4) 2024-04-29 597ff08ac1 OpenJDK 64-Bit Server VM 22.0.1+8-16 on 22.0.1+8-16 +jit [x86_64-linux]
$ uname -a
Linux PC-DE2307E1253R 5.15.146.1-microsoft-standard-WSL2 #1 SMP Thu Jan 11 04:09:03 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux
```

**Expected Behavior**
When running the tmp.rb on Ubuntu's java, the expected result is obtained.
```ruby
require 'open3'
File.write('tmp.txt', '0' * 70000)

Open3.popen3('cat tmp.txt') do |i, o, e, t|
    puts o.gets.size # 65536
end

Open3.popen3('ruby -e "print ?0 * 70000"') do |i, o, e, t|
    puts o.gets.size # 70000
end

Open3.popen3('ruby -e "print ?0 * 70000"|cat') do |i, o, e, t|
    puts o.gets.size # 65536
end
```

```
$ java --version
openjdk 21.0.3 2024-04-16
OpenJDK Runtime Environment (build 21.0.3+9-Ubuntu-1ubuntu1)
OpenJDK 64-Bit Server VM (build 21.0.3+9-Ubuntu-1ubuntu1, mixed mode, sharing)
$ java -jar jruby-complete-9.4.7.0.jar tmp.rb
70000
70000
70000
```

**Actual Behavior**
However, when running with openjdk 21 or 22, the output of cat is limited to 65536 characters.
(Strangely, the output from the second ruby command is as expected.)
```
$ ./jdk-22.0.1/bin/java --version
openjdk 22.0.1 2024-04-16
OpenJDK Runtime Environment (build 22.0.1+8-16)
OpenJDK 64-Bit Server VM (build 22.0.1+8-16, mixed mode, sharing)
$ ./jdk-22.0.1/bin/java -jar jruby-complete-9.4.7.0.jar tmp.rb
65536
70000
65536
```

**Comment by headius** [comment]
Wow that is a weird one. I bet it relates to #8069.

**Comment by yamam** [comment]
It worked properly in JDK 23.
https://download.java.net/java/early_access/jdk23/22/GPL/openjdk-23-ea+22_linux-x64_bin.tar.gz

```
$ ./jdk-23/bin/java -jar jruby-complete-9.4.7.0.jar tmp.rb
70000
70000
70000
```

**Comment by ntkme** [comment]
This affects users of `sass-embedded` gem. This gem runs a dart vm child process for `dart-sass`, and uses stdin/stdout pipe to stream protobuf messages for communication.

When user compiles a sass input that has a very large output, where the encoded protobuf message size goes beyond 65536 bytes, the `dart-sass` child process would permanently get stuck with `OS Error: Resource temporarily unavailable, errno = 35` trying to write to stdout.

See https://github.com/sass-contrib/sass-embedded-host-ruby/issues/208

**Comment by chadlwilson** [comment]
I'd like to sanity test with JDK 23 also, however our whole build infra relies on Gradle which doesn't yet allow running with Java 23, so will require a bit more hacking to sanity test just this piece.

If the root cause is the same as #8069 then perhaps we need to wait for the `21.0.4` backport in https://bugs.openjdk.org/browse/JDK-8329604 ?

**Comment by ntkme** [comment]
@chadlwilson I have confirmed that JDK 23-ea fixed this issue for sass-embedded:

https://github.com/sass-contrib/sass-embedded-host-ruby/actions/runs/9074470469/job/24933277846

**Comment by headius** [comment]
> If the root cause is the same as #8069 then perhaps we need to wait for the `21.0.4` backport in https://bugs.openjdk.org/browse/JDK-8329604 in July 2024?

With @ntkme confirming it is fixed in an early-access build of 23, I suspect this was the right fix.

**Comment by headius** [comment]
Workarounds described in #8069 to force the streams to non-blocking in the child process. As of 9.4.7.0 this code should fix it:

require 'fcntl'
flags = write.fcntl(Fcntl::F_GETFL)
write.fcntl(Fcntl::F_SETFL, flags | (Fcntl::O_NONBLOCK)) # set nonblock since we don't know it's nonblock in JRuby
write.fcntl(Fcntl::F_SETFL, flags & (~Fcntl::O_NONBLOCK)) # clear nonblock
```

**Comment by headius** [comment]
Marking as duplicate of #8069, fixed in JDK 21.0.4 and 23+ by https://bugs.openjdk.org/browse/JDK-8329604.

**Comment by chadlwilson** [comment]
Ok, let me try the workaround for now, as the issue here seems to be only at build time for us, so would be nice to move forward.
